External/Plugins/FlashDebugger/Controls/StackframeUI.cs

```diff
@@ -40,81 +166,346 @@ public StackframeUI(PluginMain pluginMain, ImageList imageList)
+                }
+            }
+
+            bool lastExternal = false;
+            int currentFrame = PluginMain.debugManager.CurrentFrame;
+            foreach (var item in wholeFrameStack)
+            {
+                bool match = true;
+                item.ImageIndex = -1;
+                if (filterText != string.Empty)
+                {
+                    if (regex != null)
+                    {
+                        match = regex.IsMatch(item.SubItems[1].Text);
+                    }
+                    else
+                    {
+                        match = item.SubItems[1].Text.IndexOf(filterText,
+                            toolStripItemMatchCase.Checked ? StringComparison.Ordinal
+                                : StringComparison.OrdinalIgnoreCase) > -1;
+                    }
+
+                    if (toolStripItemNegate.Checked) match = !match;
+                }
+
+                if (match)
+                {
+                    // TODO: Check proper theming colour
+                    if (justMyCode && item.SubItems[1].ForeColor == System.Drawing.SystemColors.GrayText)
+                    {
+                        if (lastExternal)
+                        {
+                            if (((ListItemData) item.Tag).Index == currentFrame)
+                            {
+                                lv.Items[lv.Items.Count - 1].ImageIndex = currentImageIndex;
+                                lastSelected = lv.Items.Count - 1;
+                            }
+                            continue;
+                        }
+                        var newItem = lv.Items.Add(new ListViewItem(new[] { string.Empty, "[External Code]" }, -1)
+                        {
+                            Tag = new ListItemData { Index = -1 }
+                        });
+
+                        newItem.UseItemStyleForSubItems = false;
+                        // TODO: Apply proper theming colour
+                        newItem.SubItems[1].ForeColor = System.Drawing.SystemColors.GrayText;
+                        if (((ListItemData) item.Tag).Index == currentFrame)
+                        {
+                            newItem.ImageIndex = currentImageIndex;
+                            lastSelected = lv.Items.Count - 1;
+                        }
+                        
+                        lastExternal = true;
+                    }
+                    else
+                    {
+                        lastExternal = false;
+                        lv.Items.Add(item);
+                        if (((ListItemData)item.Tag).Index == currentFrame)
+                        {
+                            item.ImageIndex = currentImageIndex;
+                            lastSelected = lv.Items.Count - 1;
+                        }
+                    }
+                }
+            }
+
+            wholeFrameStack[PluginMain.debugManager.CurrentFrame].ImageIndex = currentImageIndex;
```

Comment:
At least as code currently stands there won't be indexes out of range, but I'll make more tests.